src/meta/src/rpc/service/ddl_service.rs

```diff
@@ -604,36 +607,55 @@ where
             .start_create_stream_job_procedure(stream_job)
             .await?;
 
-        // 6. Build actor graph from the fragment graph.
-        // TODO: directly store the freezed `stream_job`.
-        let mut ctx = CreateStreamingJobContext {
-            streaming_definition: stream_job.mview_definition(),
-            table_properties: stream_job.properties(),
-            table_mview_map: self
-                .fragment_manager
-                .get_build_graph_info(dependent_relations)
-                .await?
-                .table_mview_actor_ids,
-            dependent_table_ids: dependent_relations.clone(),
-            internal_tables,
-            ..Default::default()
-        };
-
+        // 6. Resolve the upstream fragments, extend the fragment graph to a complete graph that
+        // contains all information needed for building the actor graph.
         let upstream_mview_fragments = self
             .fragment_manager
-            .get_upstream_mview_fragments(dependent_relations)
+            .get_upstream_mview_fragments(&dependent_relations)
             .await?;
+        let upstream_mview_actors = upstream_mview_fragments
+            .iter()
+            .map(|(&table_id, fragment)| {
+                (
+                    table_id,
+                    fragment.actors.iter().map(|a| a.actor_id).collect_vec(),
+                )
+            })
+            .collect();
+
         let complete_graph =
             CompleteStreamFragmentGraph::new(fragment_graph, upstream_mview_fragments)?;
 
-        // TODO(bugen): we should merge this step with the `Scheduler`.
-        let actor_graph_builder = ActorGraphBuilder::new(complete_graph, default_parallelism)?;
-
-        let graph = actor_graph_builder
-            .generate_graph(self.env.id_gen_manager_ref(), &mut ctx)
+        // 7. Build the actor graph.
+        let cluster_info = self.cluster_manager.get_streaming_cluster_info().await;
+        let actor_graph_builder =
+            ActorGraphBuilder::new(complete_graph, cluster_info, default_parallelism)?;
+
+        let ActorGraphBuildResult {
+            graph,
+            building_locations,
+            existing_locations,
+            dispatchers,
+        } = actor_graph_builder
+            .generate_graph(self.env.id_gen_manager_ref(), stream_job)
             .await?;
 
-        // 7. mark creating tables, including internal tables and the table of the stream job.
+        // 8. Build the table fragments structure that will be persisted in the stream manager, and
+        // the context that contains all information needed for building the actors on the compute
+        // nodes.
+        let table_fragments =
+            TableFragments::new(id.into(), graph, &building_locations.actor_locations, env);
+
+        let ctx = CreateStreamingJobContext {
```

Comment:
The `Context` is now immutable once created. It's majorly used to pass multiple arguments.